### ISSUE TITLE
Investigate JITServer monitor usage

### ISSUE THREAD

**Issue by dmitry-ten** [eclipse/openj9#11596]
The more clients are connected to a single server, the slower the compilation ramp-up is happening, even when the server has more than enough CPU and memory resources available. This scaling issue could be partially caused by high contestion of global server-side monitors. We need to investigate which monitors are the most hotly contested and whether there is anything that can be done to improve scaling.

**Comment by dmitry-ten** [comment]
I ran JITServer with JLM (Java Lock Monitor) tracking enabled. I connected 9 clients using 1 CPU to the server, waited for them to start up, then applied load using JMeter to each client. This experiment makes all clients request lots of remote compilations simultaneously, which puts a very large load on the server. The server has been allocated 26 CPUs and 8GB of RAM, so it has plenty of resources and never comes close to exhausting them during the experiment.

These are the results for the most used monitors on JITServer:
 ```
System (Registered) Monitors
 %MISS     GETS   NONREC     SLOW      REC    TIER2    TIER3 %UTIL AVER_HTM  MON-NAME
    27   121376    91528    25008    29848  4011835   122302     5   448698  [00007F3EB561E068] JIT-CompilationQueueMonitor
     3       70       69        2        1      571       17     0   807766  [00000000021C32F8] VM thread list
     0  1478940  1478940     5483        0 10156607   310432     1     4822  [00007F3EB561DAE8] JIT-MemoryAllocMonitor
     0    29891    29891      106        0   387245    11569     0   115212  [00007F3EB56FA828] JIT-VerboseLogMonitor
     0     4427     4427        8        0    16216      497     0    24580  [00007F3E8C00A5E8] JIT-JITServerStaticMapMonitor
     0   478353   475782      689     2571  1503415    45942     1    12197  [00007F3E8C00A118] JIT-JITServerCHTableMonitor
     0   456225   453717      644     2508  1407213    43011     1    12554  [00007F3E8C00A8A8] JIT-JITServerCHTableMonitor
     0  4952253  4462056     5945   490197 13147576   398964     3     5927  [00007F3EB56FBED8] JIT-JITServerROMMapMonitor
     0   435334   432885      568     2449  1339867    40942     0     9017  [00007F3EB56FC508] JIT-JITServerCHTableMonitor
     0  5139878  4612448     5993   527430 13357365   405452     3     6013  [00007F3E8C00A278] JIT-JITServerROMMapMonitor
     0   356030   356030      461        0   846882    25858     0     4355  [00007F3E8C00B458] JIT-JITServerClassMapMonitor
     0  4919676  4428581     5678   491095 12859943   388984     3     6155  [00007F3E8C00BB38] JIT-JITServerROMMapMonitor
     0     4713     4713        6        0    11026      337     0    31898  [00007F3EB56FCA88] JIT-JITServerStaticMapMonitor
     0   475343   472777      600     2566  1346792    41154     1    11868  [00007F3E8C009988] JIT-JITServerCHTableMonitor
     0    13694    13694       17        0    26866      821     0    22470  [00007F3E8C00BF58] JIT-JITServerThunkSetMonitor
     0   349838   349838      431        0   783158    23925     0     4576  [00007F3E8C00BBE8] JIT-JITServerClassMapMonitor
     0  5321472  4772708     5828   548764 13113176   397712     3     5590  [00007F3EB56FCEA8] JIT-JITServerROMMapMonitor
     0  5007598  4494282     5480   513316 12632749   381834     3     5511  [00007F3E8C00B3A8] JIT-JITServerROMMapMonitor
     0  5381381  4838818     5739   542563 12764885   386928     3     5036  [00007F3E8C009AE8] JIT-JITServerROMMapMonitor
     0   352877   352877      408        0   784155    23966     0     4319  [00007F3EB56FBF88] JIT-JITServerClassMapMonitor
     0   366324   366324      417        0   763852    23337     0     4232  [00007F3E8C00A328] JIT-JITServerClassMapMonitor
     0   382488   382488      430        0   770606    23544     0     4395  [00007F3E8C009B98] JIT-JITServerClassMapMonitor
     0   383592   383592      414        0   748777    22871     0     4204  [00007F3E8C009408] JIT-JITServerClassMapMonitor
     0    15418    15418       16        0    34333     1052     0    32913  [00007F3EB56FCB38] JIT-JITServerThunkSetMonitor
     0   405687   405687      416        0   717940    21937     0     4214  [00007F3EB56FB7F8] JIT-JITServerClassMapMonitor
     0   442519   440092      436     2427  1094591    33431     1    11450  [00007F3E8C00B9D8] JIT-JITServerCHTableMonitor
     0   505086   502459      497     2627  1206908    36890     1    10587  [00007F3EB56FBD78] JIT-JITServerCHTableMonitor
     0   433962   431561      395     2401  1072695    32795     0     6868  [00007F3E6C0097A8] JIT-JITServerCHTableMonitor
     0   482261   479670      432     2591  1076524    32885     1    11061  [00007F3EB56FCD48] JIT-JITServerCHTableMonitor
     0  5452850  4894440     4229   558410  9518211   288158     3     4942  [00007F3EB56FC718] JIT-JITServerROMMapMonitor
     0  5732355  5146778     4417   585577 10193881   307959     3     4966  [00007F3EB56FB748] JIT-JITServerROMMapMonitor
     0     4762     4762        4        0     6816      208     0    29655  [00007F3E8C009E58] JIT-JITServerStaticMapMonitor
     0    14288    14288       12        0    17676      540     0    24539  [00007F3E8C00B7C8] JIT-JITServerThunkSetMonitor
     0    14482    14482       12        0    22236      680     0    28364  [00007F3E8C00A698] JIT-JITServerThunkSetMonitor
     0     4848     4848        4        0     5892      180     0    36050  [00007F3E8C0096C8] JIT-JITServerStaticMapMonitor
     0    14841    14841       12        0    24550      751     0    30588  [00007F3E8C009F08] JIT-JITServerThunkSetMonitor
     0   387888   387888      281        0   509038    15554     0     4332  [00007F3EB56FC7C8] JIT-JITServerClassMapMonitor
     0    14713    14713        9        0    17907      548     0    25499  [00007F3E8C009778] JIT-JITServerThunkSetMonitor
     0    60734    60734       37        0    90892     2772     1   206388  [00007F3EB561CA68] VM mem segment list
     0     5010     5010        3        0     6246      192     0    22712  [00007F3EB56FBAB8] JIT-JITServerStaticMapMonitor
     0   582495   579597      341     2898   974817    29747     1     9381  [00007F3E8C00B038] JIT-JITServerCHTableMonitor
     0   480494   480494      282        0   501827    15327     0     3442  [00007F3E8C00AAB8] JIT-JITServerClassMapMonitor
     0    13690    13690        8        0    13496      412     0    27344  [00007F3EB56FC2F8] JIT-JITServerThunkSetMonitor
     0     6000     6000        3        0     4419      135     0    20117  [00007F3E8C00AD78] JIT-JITServerStaticMapMonitor
     0  6806447  6101436     2922   705011  6908588   208230     3     3793  [00007F3E8C00AA08] JIT-JITServerROMMapMonitor
     0     4410     4410        2        0     2946       90     0    42120  [00007F3EB56FC248] JIT-JITServerStaticMapMonitor
     0    15926    15926        7        0    12828      393     0    22423  [00007F3EB56FBB68] JIT-JITServerThunkSetMonitor
     0    17547    17547        7        0    13504      414     0    16467  [00007F3E8C00AE28] JIT-JITServerThunkSetMonitor
     0     4224     4224        1        0     2828       87     0    26986  [00007F3E8C00B718] JIT-JITServerStaticMapMonitor
     0     5583     5583        1        0     1473       45     0    10307  [00007F3EB561F248] JIT-CompThreadMonitor-??
     0    29952    29952        4        0     6054      185     0     3184  [00007F3EB561E698] JIT-CodeCacheListMutex
     0    59842    59842        4        0     8227      251     0     1668  [00007F3EB561E8A8] JIT-DataCacheManagerMutex
     0    90013    90013        1        0     1473       45     0      178  [00007F3EB561E5E8] CodeCacheRepositoryMonitor
     0  1270266  1146840        9   123426    14911      454     0      137  [00007F3DE0002A88] ValueProfilingMutex
     0  3271371  3271371        7        0    17446      528     0       81  [00000000021C3198] JIT-SmallBlockListsMonitor
     0   119798   119798        0        0     1848       50     0      352  [00007F3EB56FA988] management fields lock
```

The most relevant column for us is `SLOW`, which shows the number of times a thread blocked during monitor acquisition.
Another relevant statistic could be `AVER_HTM`, which shows the average hold time of a monitor. High hold time could indicate that we are holding it longer than necessary, e.g. sending remote messages while holding a monitor.

**Comment by dmitry-ten** [comment]
The most contested global monitor is `JIT-CompilationQueueMonitor`, with 25k slow acquisitions and high hold time. This monitor is used to add/remove methods to/from the compilation queue.

Another contested global monitor is `JIT-MemoryAllocMonitor` which is used by persistent memory allocator but compared to the queue monitor and per-client monitors its impact should be minimal.
 
The rest of contested monitors are per-client and control access to JITServer data caches, such as `JITServerROMMapMonitor` and ` JIT-JITServerCHTableMonitor`.

There are a couple of  things we can do to investigate the monitor usage:
1. Verify that global monitors are used only for their specific purpose, i.e. check that queue monitor is only used for managing the compilation queue and we don't hold it when not strictly necessary.
2. Check for instances of monitors being held when sending a remote message. Since messages can take indefinite amount of time to be sent and replied to, holding a monitor while waiting for a response from a client is bad.

**Comment by dmitry-ten** [comment]
@mpirvu fyi.

**Comment by dmitry-ten** [comment]
I checked usages of `JIT-CompilationQueueMonitor` and couldn't find any places it is used where it's not supposed to. I believe it's contested so much simply because we can have dozens of threads constantly sending compilation requests to a single thread on the server, and we need to acquire the monitor for each request. For reference, even with just one client and 7 compilation threads, this is the most contested lock on the server, with %MISS rate of around 5%. So %MISS rate of 27% for 9 clients should not be that surprising.
I also checked usages of `JITServerCHTableMonitor` and I don't see it held while sending messages anywhere in the code.
